memgraph-graphRAG/graphRAG.ipynb

```diff
@@ -0,0 +1,797 @@
+    "\n",
+    "1. Connect to the database \n",
+    "2. Load the .env file with the `OPENAI_API_KEY=` defined\n",
+    "3. Compute and store the node embeddings \n",
+    "4. Compute the question embedding based on key information \n",
+    "5. Perform the vector search to find the most semantically similar node\n",
+    "6. Get the relevant data that is a few hops away from the pivot node\n",
+    "7. Ask LLM the question with the relevant data "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Getting a Graph RAG answer \n",
+    "\n",
+    "def main():\n",
+    "\n",
+    "    # Create a Neo4j driver\n",
+    "    driver = neo4j.GraphDatabase.driver(\"bolt://100.64.149.141:7687\", auth=(\"\", \"\"))\n",
+    "\n",
+    "    # Load .env file\n",
+    "    load_dotenv()\n",
+    "    os.environ[\"TOKENIZERS_PARALLELISM\"] = \"false\"\n",
+    "\n",
+    "    # Load the SentenceTransformer model\n",
+    "    model = SentenceTransformer(\"paraphrase-MiniLM-L6-v2\")\n",
+    "\n",
+    "    # Compute embeddings for all nodes in the graph\n",
+    "    compute_bigger_embeddings_based_on_node(\n",
+    "        driver, model\n",
+    "    )\n",
+    "    client = AsyncOpenAI()\n",
+    "\n",
+    "    # Ask a question  (feel free to change the question) \n",
+    "    question = \"In which episode was Viserys Targaryen killed?\"\n",
+    "\n",
+    "    # Key information from the question \n",
+    "    prompt = question_prompt(question)\n",
+    "    response = asyncio.run(get_response(client, prompt))\n",
+    "    print(response)\n",
+    "    key_information = response.split(\"Key Information: \")[1].strip()\n",
+    "\n",
+    "    # Compute the embedding for the key information\n",
+    "    question_embedding = model.encode(key_information)\n",
+    "\n",
+    "    # Find the most similar node to the question embedding\n",
+    "    node = find_most_similar_node(driver, question_embedding)\n",
+    "    if node:\n",
+    "        print(\"The most similar node is:\")\n",
+    "        print(node)\n",
+    "\n",
+    "    # Get the relevant data based on the most similar node\n",
+    "    relevant_data = get_relevant_data(driver, node, hops=2)\n",
+    "\n",
+    "    # Show the relevant data\n",
+    "    print(\"The relevant data is:\")\n",
+    "    print(relevant_data)\n",
+    "\n",
+    "    # LLM answers the question based on the relevant data\n",
+    "    prompt = RAG_prompt(question, relevant_data)\n",
+    "    response = asyncio.run(get_response(client, prompt))\n",
+    "    print(\"The response is:\")\n",
+    "    print(response)\n",
+    "\n",
+    "    driver.close()\n",
+    "\n",
+    "\n",
+    "if __name__ == \"__main__\":\n",
+    "    main()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "Here are a few examples of questions and answers: \n",
```

Comment:
In the examples below, it's a bit unclear which part represents the response from the GraphRAG and where you start writing further comments. Suggestion - put the answer in this format:

> This is a nice format.